src/nmodl/nocpout.cpp

```diff
@@ -300,17 +309,22 @@ void parout() {
 	/*SUPPRESS 765*/\n\
 	");
     Lappendstr(defs_list, "extern double *hoc_getarg(int);\n");
-    if (vectorize) {
-        Sprintf(buf, "/* Thread safe. No static _p or _ppvar. */\n");
-    } else {
-        Sprintf(buf, "static double *_p; static Datum *_ppvar;\n");
-    }
-    Lappendstr(defs_list, buf);
 
     nrndeclare();
     varcount = parraycount = 0;
     declare_p();
+    // iondef defined _nrn_mechanism_cache_range
     ioncount = iondef(&pointercount); /* first is _nd_area if point process */
+    if (vectorize) {
+        Sprintf(buf, "/* Thread safe. No static _ml, _iml or _ppvar. */\n");
+    } else {
+        Sprintf(buf,
+                "static _nrn_mechanism_cache_instance _ml_real{nullptr};\n"
+                "static _nrn_mechanism_cache_range *_ml{&_ml_real};\n"
```

Comment:
`note to myself`: by looking at the generated cpp file (hh.cpp), I haven't understood the rational/use of `_nrn_mechanism_cache_instance` yet.

Reply:
In essence:
- `_nrn_mechanism_cache_instance` is something you construct from a single mechanism instance (`Prop*`), it doesn't require that the model data are sorted or that extra data produced as part of the sorting is available. This means that it has to hold/own that extra data internally for the single mechanism instance that it refers to.
- `_nrn_mechanism_cache_range` is something you construct from a `Memb_list`, i.e. a contiguous range of mechanism instances. 

The former is a subclass of the latter, so in the generated code we mostly just use `_nrn_mechanism_cache_range*` where we used to use `Memb_list*`, and it varies whether or not it's actually pointing to a `_nrn_mechanism_cache_instance`.

See https://github.com/neuronsimulator/nrn/blob/5369b161fd9f83fc3dca115636542d031b889f6f/src/neuron/cache/mechanism_range.hpp for the details.